### PR TITLE
Updates to profile

### LINKED ISSUE #891: [Feature]: automatically avoid profile error "At least one parameter needs to be estimated in phase 1"
### Describe the feature request.

@maitepons ran into, and then solved, a common problem where a profile over log(R0) fixes the only parameter in phase 1, causing an error.

I think we could make `profile()` slightly smarter, so instead of this warning about the issue (which should be an error anyway, so you have to deal with it), the function just changes the phase of the first parameter in phase 2 (or whatever positive phase is lowest) to be in phase 1. A warning could be added to say which parameter was changed such that if the user wants to modify a different one they could do so manually.

https://github.com/r4ss/r4ss/blob/43f55e12558d533c6b6ca9872032bb0e05bd88b6/R/profile.R#L437-L446

If anyone has or concerns, feel free to chime in. If I don't hear objections, I'll just make the change when I get around to it.

### Describe alternatives you have considered

Keep the status-quo.

### Statistical validity, if applicable

_No response_

### Describe if this is needed for a management application

_No response_

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/r4ss/r4ss#code-of-conduct)

### PR DESCRIPTION
**Pull request by okenk**
* Add ability to run profiles in parallel using {future} and {furrr}. Addresses #775
* Deprecate `read_like` in `profile()`
* Various other clean-up/bug fixes of `profile()`
* Add {purrr} and {future} to Imports. They were already transitively imported via {furrr}, this just eliminates some warnings when checking the package.
* Adjust warning suppression of `SS_parlines()` to work with {future}.
* Change examples in `jitter()` and `retro()` to use {parallelly} instead of {parallel}.
* Change first alphabetical phase 2 parameter to phase 1 if profile results in no phase 1 parameters. Closes #891.

### CODE REVIEW

**Comment by okenk** [comment]
As I was waiting for profiles to run quite slowly, I was wondering the status of this! Thanks!

**Review by iantaylor-NOAA — CHANGES_REQUESTED** [review]
@okenk, thanks for all your work parallelizing r4ss functions and sorry for letting this PR languish. At first I wanted to make sure we resolved the issues with parallelizing `jitter()` but then forgot about this one after we got that function working better.

The changes look good in general. I'm suggesting a few minor changes and a few of the comments that were really questions can be now removed.

The one significant issue is that while running the example just trying to function I ran into an error:

```
Changing working directory to C:\Users\IAN~1.TAY\AppData\Local\Temp\RtmpuAtIHN/profile,
 but will be changed back on exit from function.
running profile i=1/7
Error in (function (.x, .f, ..., .progress = FALSE)  : 
  i In index: 1.
Caused by error in `file()`:
! cannot open the connection
In addition: Warning messages:
1: In copy_SS_inputs(dir.old = getwd(), dir.new = profile_dir, overwrite = TRUE,  :
  file not found: C:/github/r4ss/starter.ss
2: In file(con, "r") :
  cannot open file 'C:/github/r4ss/control.ss': No such file or directory
```

This seems like it's related to the `setwd()` not working correctly within the parallel processes. Indeed, I'm seeing the `profile1`...`profile7` folders getting written within my original working directory. But I'm struggling to debug, in part because I have to use `devtools::install()` (which takes a long time) rather than `devtools::load_all()` as discussed in https://github.com/r4ss/r4ss/pull/909#pullrequestreview-1866684665.

File: R/profile.R [hunk 1]
```diff
@@ -397,25 +414,40 @@ profile <- function(dir,
       " profile(..., usepar = TRUE)."
     )
   }
-
+  # check for consistency of par settings and future settings
+  if(usepar & !globalpar & !is(future::plan(), 'sequential')) {
+    message(
+      "usepar = TRUE and globalpar = FALSE, but you are attempting to run\n",
+      "the profile in parallel. Changing future strategy to sequential.\n",
+      "It will return to your original strategy upon exit."
```
[the remark this answers is not in the document]
Comment (okenk) [reply R/profile.R:422]: Yes, I think both a) running jitters, retrospectives, and parallels in one fell swoop (e.g., if you use one call to nwfscDiags there would be no way to change your parallel strategy) and also just b) generally returning things as you found them.

File: R/profile.R [hunk 2]
```diff
@@ -397,25 +414,40 @@ profile <- function(dir,
...
+    )
+    # save current strategy as oplan, and change it to sequential, all in one step!
+    oplan <- future::plan(future::sequential) 
+    on.exit(future::plan(oplan), add = TRUE)
+  }
+  
   # back up par file
   if (usepar) {
     file.copy("ss.par", "parfile_original_backup.sso")
```
[the remark this answers is not in the document]
Comment (okenk) [reply R/profile.R:431]: There were some non-trivial merge conflicts from fbe6f5270d0acef11114bb517722e7363629f6ab, also related to par file naming. How do you suggest I address those? Should I use the object `parfile` or hardcode in ss3.par with a warning?

[the remark this answers is not in the document]
Comment (okenk) [reply R/profile.R:431]: (Non-trivial because I have messed with the file so much.)

[the remark this answers is not in the document]
Comment (iantaylor-NOAA) [reply R/profile.R:431]: I forgot that there were already changes made to the function. Using `parfile` makes sense to me. I meant to say that I'm happy to help with any of these edits next week if you want help, including merging those par-file-related changes.

**Comment by okenk** [comment]
It appears that when you are running things in parallel, the `on.exit()` is activated, and it returns your working directory back to its original state. 

The following code works if my plan is sequential, but I get essentially the same error as you if the plan is multisession:
```
test.func <- function(dir) {
  orig_wd <- getwd()
  on.exit(setwd(orig_wd))
  setwd(dir)
  
  furrr::future_map(1:2, \(x) {
    r4ss::SS_read(dir = getwd())
  })
}
xx <- test.func('inst/extdata/simple_small')
```

It looks like you can reset the working directory before the call to `r4ss::SS_read()` and that should eliminate the error. The file management has easily been the trickiest part of getting this set up, at least for me. Weird and unpredictable things seem to happen...

I *think* this also means if `profile()` changed your plan to sequential, it will change the plan back to whatever it was when you started. I will try to investigate if this matters or not. (I tend to think it won't.)

**Comment by e-perl-NOAA** [comment]
I looked into this a bit more with a [version of the jitter function](https://github.com/r4ss/r4ss/blob/try_fix_jitter_dir_issue/R/jitter.R) that I have been working on and when using multisession (before the jitter function like in the example), on.exit() didn't bring me back to the right working directory which was causing issues. However, when I added
```
ncores <- parallelly::availableCores(omit = 1)
future::plan(future::multisession, workers = ncores)
```
inside of the function, the on.exit command worked to bring me back to the initial directory. I did this after searching on github for other examples of code that had "on.exit(setwd(" and "multisession" somewhere in their R code. 

In my branch, I also changed out where it defaults to getwd() in a lot of places to specifying a directory because the situation in which I want to apply the jitter function, the directory that I am working in is not where I want to run the jitter function, I want to run it in a subdirectory.

Also, just to fill you in on some of the other changes I made in my branch, the "..." syntax in the functions confused me because I there were a lot of times that I couldn't remember what the key params in that function were, so I added some of those params back for clarity's sake on my part.

**Comment by e-perl-NOAA** [comment]
They are just changes I've been experimenting with trying to figure out what is going on and potential fixes.

**Comment by e-perl-NOAA** [comment]
I agree @okenk I think avoiding setting working directories in functions (outside of if a directory is needed and none is provided to set the directory as the current working directory) is the move. I made changes in my branch to reflect this (again, except for when no dir is provided to set the dir as the current working directory), and removed where I set the future in the jitter function. I also added the dir provided to several places where one was needed to get the right file (specifically SS_readstarter(), SS_writestarter(), and some file copy functions). The function in the branch seems to be working for the very simple example that I have tested it with. Idk if you would like to take the changes that you like (if any) and add them to the changes that you are working on in your fork.

**Comment by okenk** [comment]
These sound great @e-perl-NOAA! I will take a look at them. I may not get to this today.

**Review by iantaylor-NOAA — APPROVED** [review]
Changes good and profile ran in parallel without issue on my computer.
The simple_small model took about 2 min 30 sec for each profile step and all 7 were done in less than 3 minutes.
Thank you @okenk!

**Pull request merged**